File: apps/app/src/api/generation.ts
import httpClient from "@/lib/http";
import type { GenerationKind } from "@/lib/stores/generation-draft";

export type { GenerationKind };

export interface GenerationParamOption {
	label: string;
	value: string;
}

export interface GenerationParam {
	name: string;
	label: string;
	type: "select" | "number" | "boolean" | "text";
	default?: unknown;
	options?: GenerationParamOption[];
	required?: boolean;
	min?: number;
	max?: number;
	help?: string;
}

export type GenerationProvider =
	| "openai"
	| "google"
	| "volcengine"
	| "dmx"
	| "openrouter"
	| "jimeng";
export type GenerationProviderType = "official" | "aggregator" | "local";
export type GenerationRouteStatus = "available" | "planned" | "gated";

export interface GenerationProviderInfo {
	id: GenerationProvider;
	label: string;
	providerType: GenerationProviderType;
}

export interface GenerationRoute {
	id: string;
	familyId: string;
	versionId: string;
	label: string;
	kind: GenerationKind;
	provider: GenerationProvider;
	model: string;
	adapter: string;
	docUrl: string;
	async: boolean;
	supportsReferenceUrls: boolean;
	status: GenerationRouteStatus;
	statusReason?: string;
	params: GenerationParam[];
	legacyModelId?: string;
	configured?: boolean;
}

export interface GenerationFamily {
	id: string;
	label: string;
	kind: GenerationKind;
	description?: string;
}

export interface GenerationVersion {
	id: string;
	familyId: string;
	label: string;
	kind: GenerationKind;
	canonicalModel: string;
	capabilities: {
		async: boolean;
		supportsReferenceUrls: boolean;
	};
}

export interface GenerationModel {
	id: string;
	label: string;
	kind: GenerationKind;
	provider: string;
	model: string;
	adapter: string;
	docUrl: string;
	async: boolean;
	supportsReferenceUrls: boolean;
	params: GenerationParam[];
}

export interface GenerationModelsResponse {
	families: GenerationFamily[];
	versions: GenerationVersion[];
	routes: GenerationRoute[];
	models: GenerationModel[];
	providers: GenerationProviderInfo[];
}

export interface GenerationAsset {
	kind: GenerationKind;
	url?: string;
	base64?: string;
	mimeType?: string;
}

export interface GenerationUsage {
	inputTokens: number;
	outputTokens: number;
	totalTokens: number;
}

export interface GenerationMessageRequest {
	kind: GenerationKind;
	conversationId?: string;
	sessionId?: string;
	scopeId?: string;
	routeId: string;
	familyId?: string;
	versionId?: string;
	provider?: GenerationProvider;
	modelId: string;
	model: string;
	prompt: string;
	referenceUrls: string[];
	referenceAssetIds: string[];
	params: Record<string, unknown>;
}

export interface GenerationMessageResponse {
	id: string;
	role: "assistant";
	status: string;
	message: string;
	assets: GenerationAsset[];
	usage: GenerationUsage;
	error?: string;
}

export interface GenerationTask {
	id: string;
	conversationId?: string;
	sessionId?: string;
	kind: GenerationKind;
	routeId: string;
	familyId: string;
	versionId: string;
	provider: GenerationProvider;
	modelId: string;
	model: string;
	prompt: string;
	referenceUrls: string[];
	referenceAssetIds: string[];
	params: Record<string, unknown>;
	status: string;
	message: string;
	assets: GenerationAsset[];
	usage: GenerationUsage;
	error?: string;
	createdAt: string;
	updatedAt: string;
	durationMs?: number;
	attempts?: GenerationTaskAttempt[];
	retryCount: number;
	lastAttemptAt?: string;
}

export interface GenerationTaskAttempt {
	id: string;
	taskId: string;
	action: string;
	status: string;
	message?: string;
	error?: string;
	createdAt: string;
}

export interface GenerationTasksResponse {
	tasks: GenerationTask[];
}

export interface GenerationConversation {
	id: string;
	sessionId?: string;
	scopeId: string;
	kind: GenerationKind;
	title: string;
	taskCount: number;
	latestPrompt?: string;
	createdAt: string;
	updatedAt: string;
}

export interface GenerationConversationsResponse {
	conversations: GenerationConversation[];
}

export interface CreateGenerationConversationRequest {
	sessionId?: string;
	scopeId?: string;
	kind: GenerationKind;
	title: string;
}

export const generationModelsKey = "/generation/models";
export const generationTasksKey = "/generation/tasks";
export const generationConversationsKey = "/generation/sessions";
export const defaultGenerationConversationScopeId = "studio";
const generationRequestTimeoutMs = 1_000_000;

export const getGenerationModels = async () => {
	const response = await httpClient.get<GenerationModelsResponse>(generationModelsKey);
	return response.data;
};

export const generationTasksQueryKey = (
	conversationId?: string | null,
	kind?: GenerationKind,
	scopeId = defaultGenerationConversationScopeId,
) => [generationTasksKey, scopeId, conversationId?.trim() || "", kind ?? ""] as const;

export const generationConversationsQueryKey = (
	kind?: GenerationKind,
	scopeId = defaultGenerationConversationScopeId,
) => [generationConversationsKey, scopeId, kind ?? ""] as const;

export const generationTaskQueryKey = (id: string) => [generationTasksKey, id] as const;

export const getGenerationTasks = async (
	conversationId?: string | null,
	kind?: GenerationKind,
	scopeId = defaultGenerationConversationScopeId,
) => {
	void scopeId;
	const sessionId = conversationId?.trim();
	const response = await httpClient.get<GenerationTasksResponse>(
		sessionId
			? `${generationConversationsKey}/${encodeURIComponent(sessionId)}/tasks`
			: generationTasksKey,
		{
		params: {
			...(kind ? { kind } : {}),
		},
		},
	);
	return { ...response.data, tasks: response.data.tasks.map(normalizeGenerationTask) };
};

export const getGenerationTask = async (id: string) => {
	const response = await httpClient.get<GenerationTask>(
		`${generationTasksKey}/${encodeURIComponent(id)}`,
	);
	return response.data;
};

export const getGenerationConversations = async (
	kind?: GenerationKind,
	scopeId = defaultGenerationConversationScopeId,
) => {
	void scopeId;
	const response = await httpClient.get<GenerationConversationsResponse>(
		generationConversationsKey,
		{
			params: {
				...(kind ? { kind } : {}),
			},
		},
	);
	const sessions =
		(response.data as { sessions?: GenerationConversation[]; conversations?: GenerationConversation[] })
			.sessions ??
		(response.data as { sessions?: GenerationConversation[]; conversations?: GenerationConversation[] })
			.conversations ??
		[];
	return { ...response.data, conversations: sessions.map(normalizeGenerationConversation) };
};

export const createGenerationConversation = async (
	request: CreateGenerationConversationRequest,
) => {
	const rawRequest = request as CreateGenerationConversationRequest & { id?: string };
	const response = await httpClient.post<GenerationConversation>(
		generationConversationsKey,
		{
			sessionId: request.sessionId ?? rawRequest.id,
			kind: request.kind,
			title: request.title,
		},
	);
	return normalizeGenerationConversation(response.data);
};

export const sendGenerationMessage = async (request: GenerationMessageRequest) => {
	const payload = generationMessagePayload(request);
	const response = await httpClient.post<GenerationMessageResponse>(
		`${generationConversationsKey}/${encodeURIComponent(payload.sessionId)}/messages`,
		payload,
		{
			timeout: generationRequestTimeoutMs,
		},
	);
	return response.data;
};

const generationMessagePayload = (request: GenerationMessageRequest) => {
	const sessionId =
		request.sessionId?.trim() ||
		request.conversationId?.trim() ||
		defaultGenerationConversationScopeId;
	const payload = { ...request };
	delete payload.conversationId;
	delete payload.scopeId;
	return { ...payload, sessionId };
};

const normalizeGenerationConversation = (conversation: GenerationConversation) => {
	const sessionId = conversation.sessionId ?? conversation.id;
	return {
		...conversation,
		id: sessionId,
		sessionId,
		scopeId: conversation.scopeId ?? sessionId,
	};
};

const normalizeGenerationTask = (task: GenerationTask) => {
	const sessionId = task.sessionId ?? task.conversationId ?? "";
	return {
		...task,
		sessionId,
		conversationId: sessionId,
	};
};

File: packages/server/internal/http/dto/generation.go
package dto

import coregeneration "github.com/torchstellar-team/mediago-drama/packages/core/pkg/generation"

// GenerationModelsResponse returns the generation catalog.
type GenerationModelsResponse struct {
	Families  []coregeneration.ModelFamily  `json:"families"`
	Versions  []coregeneration.ModelVersion `json:"versions"`
	Routes    []coregeneration.ModelRoute   `json:"routes"`
	Models    []coregeneration.ModelSpec    `json:"models"`
	Providers []coregeneration.ProviderInfo `json:"providers"`
}

// GenerationMessageRequest creates or retries a generation request.
type GenerationMessageRequest struct {
	Kind               string                        `json:"kind" ts:"Kind"`
	ConversationID     string                        `json:"sessionId,omitempty"`
	ScopeID            string                        `json:"-"`
	ProjectID          string                        `json:"projectId,omitempty"`
	SectionID          string                        `json:"sectionId,omitempty"`
	CapabilityID       string                        `json:"capabilityId,omitempty"`
	NotificationTarget *GenerationNotificationTarget `json:"notificationTarget,omitempty"`
	RouteID            string                        `json:"routeId"`
	FamilyID           string                        `json:"familyId,omitempty"`
	VersionID          string                        `json:"versionId,omitempty"`
	Provider           string                        `json:"provider,omitempty"`
	ModelID            string                        `json:"modelId"`
	Model              string                        `json:"model"`
	Prompt             string                        `json:"prompt"`
	Size               string                        `json:"size,omitempty"`
	ReferenceURLs      []string                      `json:"referenceUrls"`
	ReferenceAssetIDs  []string                      `json:"referenceAssetIds"`
	Params             map[string]any                `json:"params"`
}

// GenerationMessageResponse is returned by generation create/poll/retry calls.
type GenerationMessageResponse struct {
	ID        string            `json:"id"`
	Role      string            `json:"role"`
	Status    string            `json:"status"`
	Message   string            `json:"message"`
	Text      string            `json:"text,omitempty"`
	Assets    []GenerationAsset `json:"assets"`
	Usage     GenerationUsage   `json:"usage"`
	Error     string            `json:"error,omitempty"`
	ErrorCode string            `json:"errorCode,omitempty"`
	ErrorType string            `json:"errorType,omitempty"`
	Retryable bool              `json:"retryable,omitempty"`
}

// GenerationTextStreamEvent is one server-sent event for text generation.
type GenerationTextStreamEvent struct {
	Type           string                     `json:"type"`
	TaskID         string                     `json:"taskId,omitempty"`
	ConversationID string                     `json:"sessionId,omitempty"`
	Delta          string                     `json:"delta,omitempty"`
	Message        *GenerationMessageResponse `json:"message,omitempty"`
	Status         string                     `json:"status,omitempty"`
	Error          string                     `json:"error,omitempty"`
	Usage          *GenerationUsage           `json:"usage,omitempty"`
}

// GenerationNotificationSectionTarget identifies a document section.
type GenerationNotificationSectionTarget struct {
	BlockID           string `json:"blockId"`
	DocumentID        string `json:"documentId"`
	HeadingLevel      int    `json:"headingLevel"`
	HeadingOccurrence int    `json:"headingOccurrence"`
	HeadingText       string `json:"headingText"`
	Markdown          string `json:"markdown"`
	PlainText         string `json:"plainText"`
	Prompt            string `json:"prompt"`
}

// GenerationNotificationTarget identifies where a generation notification opens.
type GenerationNotificationTarget struct {
	Kind          string                              `json:"kind"`
	ProjectID     string                              `json:"projectId,omitempty"`
	DocumentID    string                              `json:"documentId,omitempty"`
	DocumentTitle string                              `json:"documentTitle,omitempty"`
	Section       GenerationNotificationSectionTarget `json:"section"`
}

// GenerationNotificationRecord is a persisted user-facing generation notification.
type GenerationNotificationRecord struct {
	ID          string                       `json:"id"`
	TaskID      string                       `json:"taskId"`
	TaskKind    string                       `json:"taskKind" ts:"Kind"`
	TaskStatus  string                       `json:"taskStatus"`
	ProjectID   string                       `json:"projectId,omitempty"`
	Title       string                       `json:"title"`
	Description string                       `json:"description"`
	AssetCount  int                          `json:"assetCount"`
	ReadAt      string                       `json:"readAt,omitempty"`
	Target      GenerationNotificationTarget `json:"target"`
	CreatedAt   string                       `json:"createdAt"`
	UpdatedAt   string                       `json:"updatedAt"`
}

// GenerationNotificationsResponse lists generation notifications.
type GenerationNotificationsResponse struct {
	Notifications []GenerationNotificationRecord `json:"notifications"`
}

// GenerationNotificationEvent is one SSE payload for generation notifications.
type GenerationNotificationEvent struct {
	ID           string                       `json:"id"`
	Type         string                       `json:"type"`
	ProjectID    string                       `json:"projectId,omitempty"`
	Notification GenerationNotificationRecord `json:"notification"`
	CreatedAt    string                       `json:"createdAt"`
}

// GenerationAsset is a generated asset reference or inline payload.
type GenerationAsset struct {
	Kind     string `json:"kind" ts:"Kind"`
	URL      string `json:"url,omitempty"`
	Base64   string `json:"base64,omitempty"`
	MIMEType string `json:"mimeType,omitempty"`
}

// GenerationUsage contains token usage for generation providers.
type GenerationUsage struct {
	InputTokens     int `json:"inputTokens"`
	OutputTokens    int `json:"outputTokens"`
	TotalTokens     int `json:"totalTokens"`
	ReasoningTokens int `json:"reasoningTokens"`
	CachedTokens    int `json:"cachedTokens"`
}

// GenerationTaskRecord is a persisted generation task.
type GenerationTaskRecord struct {
	ID                string                        `json:"id"`
	ProviderTaskID    string                        `json:"providerTaskId,omitempty"`
	ConversationID    string                        `json:"sessionId,omitempty"`
	ProjectID         string                        `json:"projectId,omitempty"`
	SectionID         string                        `json:"sectionId,omitempty"`
	CapabilityID      string                        `json:"capabilityId,omitempty"`
	Kind              string                        `json:"kind" ts:"Kind"`
	RouteID           string                        `json:"routeId"`
	FamilyID          string                        `json:"familyId"`
	VersionID         string                        `json:"versionId"`
	Provider          string                        `json:"provider"`
	ModelID           string                        `json:"modelId"`
	Model             string                        `json:"model"`
	Prompt            string                        `json:"prompt"`
	ReferenceURLs     []string                      `json:"referenceUrls"`
	ReferenceAssetIDs []string                      `json:"referenceAssetIds"`
	Params            map[string]any                `json:"params"`
	Status            string                        `json:"status"`
	Message           string                        `json:"message"`
	Text              string                        `json:"text,omitempty"`
	Assets            []GenerationAsset             `json:"assets"`
	Usage             GenerationUsage               `json:"usage"`
	Error             string                        `json:"error,omitempty"`
	ErrorCode         string                        `json:"errorCode,omitempty"`
	ErrorType         string                        `json:"errorType,omitempty"`
	Retryable         bool                          `json:"retryable,omitempty"`
	CreatedAt         string                        `json:"createdAt"`
	UpdatedAt         string                        `json:"updatedAt"`
	DurationMS        int64                         `json:"durationMs,omitempty"`
	Attempts          []GenerationTaskAttemptRecord `json:"attempts,omitempty"`
	RetryCount        int                           `json:"retryCount"`
	LastAttemptAt     string                        `json:"lastAttemptAt,omitempty"`
}

// GenerationTaskAttemptRecord is a stored task attempt.
type GenerationTaskAttemptRecord struct {
	ID        string `json:"id"`
	TaskID    string `json:"taskId"`
	Action    string `json:"action"`
	Status    string `json:"status"`
	Message   string `json:"message,omitempty"`
	Error     string `json:"error,omitempty"`
	CreatedAt string `json:"createdAt"`
}

// GenerationTasksResponse lists generation tasks.
type GenerationTasksResponse struct {
	Tasks []GenerationTaskRecord `json:"tasks"`
}

// GenerationPreferenceRecord is a persisted generation preference set.
type GenerationPreferenceRecord struct {
	SessionID     string                    `json:"sessionId"`
	ScopeID       string                    `json:"-"`
	FamilyIDs     map[string]string         `json:"familyIds" ts:"Partial<Record<Kind, string>>"`
	RouteIDs      map[string]string         `json:"routeIds"`
	VersionIDs    map[string]string         `json:"versionIds"`
	RouteParams   map[string]map[string]any `json:"routeParams"`
	StylePresetID string                    `json:"stylePresetId"`
	CreatedAt     string                    `json:"createdAt,omitempty"`
	UpdatedAt     string                    `json:"updatedAt,omitempty"`
}

// UpdateGenerationPreferenceRequest updates generation preferences for a scope.
type UpdateGenerationPreferenceRequest struct {
	ScopeID       string                    `json:"sessionId,omitempty"`
	FamilyIDs     map[string]string         `json:"familyIds" ts:"Partial<Record<Kind, string>>"`
	RouteIDs      map[string]string         `json:"routeIds"`
	VersionIDs    map[string]string         `json:"versionIds"`
	RouteParams   map[string]map[string]any `json:"routeParams"`
	StylePresetID string                    `json:"stylePresetId"`
}

// GenerationConversationRecord is a persisted generation conversation.
type GenerationConversationRecord struct {
	ID           string `json:"sessionId"`
	ScopeID      string `json:"-"`
	Kind         string `json:"kind" ts:"Kind"`
	Title        string `json:"title"`
	Default      bool   `json:"-"`
	TaskCount    int64  `json:"taskCount"`
	LatestPrompt string `json:"latestPrompt,omitempty"`
	CreatedAt    string `json:"createdAt"`
	UpdatedAt    string `json:"updatedAt"`
}

// GenerationConversationsResponse lists generation conversations.
type GenerationConversationsResponse struct {
	Conversations []GenerationConversationRecord `json:"sessions"`
}

// CreateGenerationConversationRequest creates a generation conversation.
type CreateGenerationConversationRequest struct {
	ID      string `json:"sessionId,omitempty"`
	ScopeID string `json:"-"`
	Kind    string `json:"kind" ts:"Kind"`
	Title   string `json:"title"`
}
